### PR TITLE
Chore/wss cli

### PR DESCRIPTION
**Pull request by frickjack**
Jira Ticket: [PXP-7015](https://ctds-planx.atlassian.net/browse/PXP-7015)

* use `GEN3_API_KEY` environment variable - default to `~/.gen3/credentials.json`
* derive endpoint from api key
* add access-token cache
* add access-token expiration check
* add workspace storage class and cli integration
* add auth cli integration
* arrange markdown docs
* add Jenkins smoke test

```
$ poetry run gen3 --help
Usage: gen3 [OPTIONS] COMMAND [ARGS]...

  Gen3 sdk commands

Options:
  --auth TEXT      path to api key, or shorthand for key in ~/.gen3/, or idp:
                   identifier if in workspace

  --endpoint TEXT  commons hostname - optional if API Key given
  --help           Show this message and exit.

Commands:
  auth  Gen3 sdk auth commands
  pfb   Commands for working with Portable Format for Biomedical Data (PFB)
  wss   Gen3 sdk workspace storage commands
```
```
$ poetry run gen3 auth --help
Usage: gen3 auth [OPTIONS] COMMAND [ARGS]...

  Gen3 sdk auth commands

Options:
  --help  Show this message and exit.

Commands:
  access-token  Get an access token suitable to pass as an Authorization...
  curl          Get an access token suitable to pass as an Authorization...
  endpoint      Get an access token suitable to pass as an Authorization...
  token-decode  Decode the given token file - may be "-" to indicate stdin
  wts-endpoint  list the idp's available from the wts in a Gen3 workspace...
  wts-list      list the idp's available from the wts in a Gen3 workspace...
```
```
$ poetry run gen3 wss --help
Usage: gen3 wss [OPTIONS] COMMAND [ARGS]...

  Gen3 sdk workspace storage commands

Options:
  --help  Show this message and exit.

Commands:
  cp            Upload url for the given workspace key
  download-url  Download url for the given workspace key
  ls            List the given workspace key
  rm            Remove the given workspace key
  upload-url    Upload url for the given workspace key
```

### Improvements

* use `GEN3_API_KEY` environment variable - default to `~/.gen3/credentials.json`
* derive endpoint from api key
* add access-token cache
* add access-token expiration check
* add workspace storage class and cli integration
* add auth cli integration
* arrange markdown docs
* add Jenkins smoke test

### CODE REVIEW

File: gen3/wss.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        Initialization for instance of the class to setup basic endpoint info.
+
+        Args:
+            auth_provider (Gen3Auth, optional): Gen3Auth class to handle passing your
+                token, required for admin endpoints
+        """
+        self._auth_provider = auth_provider
+
+    @backoff.on_exception(backoff.expo, Gen3CurlError, **DEFAULT_BACKOFF_SETTINGS)
+    def upload_url(self, ws, wskey):
```
Comment (paulineribeyre) [inline gen3/wss.py:72]: it's not clear to me what `ws` is, could you describe the parameters in each function docstring

File: gen3/wss.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            f.write(res.content)
+
+@backoff.on_exception(backoff.expo, Gen3CurlError, **DEFAULT_BACKOFF_SETTINGS)
+def put_url(urlstr, src_path):
+    """Simple put src_path to url with backoff"""
+    with open(src_path, 'rb') as f:
+        res = requests.put(urlstr, data=f)
+    check_curl_status(res)
+
+def trim_leading_slash(str):
```
Comment (paulineribeyre) [inline gen3/wss.py:41]: `str.lstrip("/")`

File: gen3/cli/wss.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import sys
+from gen3.wss import Gen3WsStorage, wsurl_to_tokens
+
+def clean_path(path):
+    """
+    Add ws:///@user/ prefix if necessary
+    """
+    if path[0:3] != "ws:":
+        while path and path[0] == "/":
+            path = path[1:]
```
Comment (paulineribeyre) [inline gen3/cli/wss.py:12]: `path.lstrip("/")`

File: gen3/auth.py [hunk 4]
```diff
@@ -16,70 +116,98 @@ class Gen3Auth(AuthBase):
...
-        self._endpoint = endpoint
-        self._refresh_file = refresh_file
+        # note - this is not actually a JWT refresh token - it's a 
+        #  gen3 api key with a token as the "api_key" property
         self._refresh_token = refresh_token
-        self._wts_idp = idp
         self._access_token = None
-
+        self._access_token_info = None
+        self._wts_idp = idp or "local"
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/auth.py:158]: to be consistent with the idea that:
```
refresh_file == "idp://local"
```
uses the local idp

File: gen3/wss.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        """
+        self._auth_provider = auth_provider
+
+    @backoff.on_exception(backoff.expo, Gen3CurlError, **DEFAULT_BACKOFF_SETTINGS)
+    def upload_url(self, ws, wskey):
+        """
+        Get a upload url for the given workspace key
+        """
+        wskey = trim_leading_slash(wskey)
+        res = self._auth_provider.curl("/ws-storage/upload/{}/{}".format(ws, wskey))
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/wss.py:77]: Yeah - need to make a jira for that.
THere's not much there: https://github.com/uc-cdis/ws-storage/blob/master/storage/http.go#L117

File: gen3/wss.py (hunk 1 above)
Comment (frickjack) [reply gen3/wss.py:72]: done

File: gen3/wss.py (hunk 2 above)
Comment (frickjack) [reply gen3/wss.py:41]: thanks

File: gen3/cli/wss.py (hunk 3 above)
Comment (frickjack) [reply gen3/cli/wss.py:12]: cool :+1:

File: gen3/wss.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply gen3/wss.py:76]: https://ctds-planx.atlassian.net/browse/PXP-7420

File: gen3/auth.py [hunk 6]
```diff
@@ -16,70 +116,98 @@ class Gen3Auth(AuthBase):
...
+                # most production environments are in the "default" namespace
+                # attempt to get a token from the workspace-token-service
+                self._use_wts = True
+                # hate calling a method from the constructor, but avoids copying code
+                self.get_access_token()
+            elif refresh_file:
+                try:
+                    with open(refresh_file) as f:
+                        file_data = f.read()
+                    self._refresh_token = json.loads(file_data)
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/auth.py:200]: good idea! :+1:

File: gen3/auth.py [hunk 7]
```diff
@@ -1,12 +1,112 @@
...
+
+def endpoint_from_token(token_str):
+    """
+    Extract the endpoint from a JWT issuer - ex:
+    """
+    info = decode_token(token_str)
+    urlparts = urlparse(info["iss"])
+    endpoint = urlparts.scheme + "://" + urlparts.hostname
+    if urlparts.port:
+        endpoint += str(urlparts.port)
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/auth.py:44]: seems ok

```
>>> parts = urlparse("https://whatever:33/path")
>>> parts.port
33
```

File: gen3/auth.py [hunk 8]
```diff
@@ -1,12 +1,112 @@
...
+    in the given namespace.  If idp is not set, then default to "local"
+    """
+    # attempt to get a token from the workspace-token-service
+    auth_url = get_wts_endpoint(namespace) + "/token/"
+    if idp and idp != "local":
+        auth_url += "?idp={}".format(idp)
+    resp = requests.get(auth_url)
+    return _handle_access_token_response(resp, "token")
+
+def check_curl_status(res):
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/auth.py:94]: yeah - should have just used `response.raise_for_status()` and `@backoff` on `requests.HTTPError` - somehow talked myself into using a custom exception type.  Fixing that ...

File: tests/test_auth.py [hunk 9]
```diff
@@ -30,13 +50,28 @@ def test_auth_init_in_workspace(monkeypatch):
...
+    access_token = test_key["api_key"]
+    def _mock_request(url, **kwargs):
+        assert url.endswith("/token/")
+
+        mocked_response = MagicMock(requests.Response)
+        mocked_response.status_code = 200
+        mocked_response.json.return_value = {
+            "token": access_token
+        }
+        #mocked_response.raise_for_status.side_effect = lambda *args: None
```
[the remark this answers is not in the document]
Comment (frickjack) [reply tests/test_auth.py:62]: done

File: gen3/auth.py [hunk 10]
```diff
@@ -16,70 +116,98 @@ class Gen3Auth(AuthBase):
     Automatically refreshes access tokens when they expire.
 
     Args:
-        endpoint (str, opt): The URL of the data commons. Optional if working in a Gen3 Workspace.
         refresh_file (str, opt): The file containing the downloaded JSON web token. Optional if working in a Gen3 Workspace.
+                Defaults to (env["GEN3_API_KEY"] || "credentials") if refresh_token and idp not set.
+                Includes ~/.gen3/ in search path if value does not include /.
+                Interprets "idp://wts/idp" as an idp.
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/auth.py]: for the command line integration we accept things like "idp://wts/whatever" and "accesstoken://..." - I debated whether to put the logic for that in the cli/ package or push it into the core, and decided to generalize the refresh_file argument for different credential sources

File: gen3/auth.py [hunk 11]
```diff
@@ -1,12 +1,112 @@
...
+    token_key = "access_token"
+    return _handle_access_token_response(resp, token_key)
+
+def get_wts_endpoint(namespace=os.getenv("NAMESPACE", "default")):
+    return "http://workspace-token-service.{}.svc.cluster.local".format(namespace)
+
+def get_wts_idps(namespace=os.getenv("NAMESPACE", "default")):
+    resp = requests.get(get_wts_endpoint(namespace) + "/external_oidc")
+    if resp.status_code != 200:
+        raise Exception("non-200 status {} - {}".format(resp.status_code, resp.text))
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/auth.py:79]: cool

File: gen3/cli/auth.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+@click.command()
+def wts_list():
+    """list the idp's available from the wts in a Gen3 workspace environment """
+    print(json.dumps(auth_tool.get_wts_idps(), indent=2))
+
+@click.command()
+@click.pass_context
+def info(ctx):
+    """Info about the current authentication mechanism"""
+    print(ctx.obj["auth_factory"].get().get_access_token())
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/cli/auth.py:63]: just removed this - it was not bound to the command tree

File: gen3/metadata.py [hunk 13]
```diff
@@ -19,9 +19,8 @@ class Gen3Metadata:
         This generates the Gen3Metadata class pointed at the sandbox commons while
         using the credentials.json downloaded from the commons profile page.
 
-        >>> endpoint = "https://nci-crdc-demo.datacommons.io"
-        ... auth = Gen3Auth(endpoint, refresh_file="credentials.json")
-        ... sub = Gen3Metadata(endpoint, auth)
+        >>> auth = Gen3Auth(endpoint, refresh_file="credentials.json")
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/metadata.py:22]: ok - patched metadata to work the same way as the other services with endpoint

File: gen3/auth.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply gen3/auth.py:44]: i mean when you're appending you get `endpoint = <scheme>://<hostname><port>`, not `endpoint = <scheme>://<hostname>:<port>`

File: gen3/index.py [hunk 14]
```diff
@@ -8,28 +8,28 @@
...
 
     Examples:
         This generates the Gen3Index class pointed at the sandbox commons while
         using the credentials.json downloaded from the commons profile page.
 
-        >>> endpoint = "https://nci-crdc-demo.datacommons.io"
-        ... auth = Gen3Auth(endpoint, refresh_file="credentials.json")
-        ... sub = Gen3Submission(endpoint, auth)
+        >>> auth = Gen3Auth(refresh_file="credentials.json")
+        ... sub = Gen3Submission(auth.endpoint, auth)
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/index.py:26]: updated the docstring :+1:

File: gen3/jobs.py [hunk 15]
```diff
@@ -29,23 +29,22 @@ class Gen3Jobs:
...
         """
         Initialization for instance of the class to setup basic endpoint info.
 
         Args:
-            endpoint (str): URL for a data commons that has jobs service(s) deployed
             auth_provider (Gen3Auth, optional): Gen3Auth class to handle passing your
                 token, required for admin endpoints
             service_location (str, optional): deployment location relative to the
                 endpoint provided
         """
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/jobs.py:45]: ok

File: gen3/submission.py [hunk 16]
```diff
@@ -25,28 +25,25 @@ class Gen3Submission:
...
-        ... sub = Gen3Submission(endpoint, auth)
+        >>> auth = Gen3Auth(refresh_file="credentials.json")
+        ... sub = Gen3Submission(auth)
 
     """
 
-    def __init__(self, endpoint, auth_provider):
-        self._auth_provider = auth_provider
-        self._endpoint = endpoint
+    def __init__(self, endpoint=None, auth_provider=None):
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/submission.py:39]: ok

File: gen3/cli/auth.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+@click.group()
+def auth():
+    """Gen3 sdk auth commands"""
+    pass
+
+auth.add_command(wts_endpoint, name="wts-endpoint")
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/cli/auth.py:70]: dropped info

File: gen3/auth.py [hunk 18]
```diff
@@ -125,38 +252,76 @@ def _handle_401(self, response, **kwargs):
...
-                if self._wts_idp:
-                    auth_url += "?idp={}".format(self._wts_idp)
-                resp = requests.get(auth_url)
-                err_msg = "Failed to get an access token from WTS at {}:\n{}"
-                token_key = "token"
+        return "bearer " + self.get_access_token()
+
+
+    def curl(self, path, request=None, data=None):
+        """Curl the given endpoint - ex: gen3 curl /user/user"""
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/auth.py]: ok - added some details :+1:

File: gen3/cli/__main__.py [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    def get(self):
+        """Lazy factory"""
+        if self._cache:
+            return self._cache
+        self._cache = gen3.auth.Gen3Auth(refresh_file=self.refresh_file)
+        return self._cache
+
+
+@click.group()
+@click.option("--auth", 'auth_config', default=os.getenv("GEN3_API_KEY", None), help="path to api key, or shorthand for key in ~/.gen3/, or idp: identifier if in workspace")
```
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/cli/__main__.py]: ok - added more details

File: gen3/auth.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/auth.py:37]: oh - you're right :+1:

[the remark this answers is not in the document]
Comment (frickjack) [reply gen3/auth.py]: ok - added the ":" - thanks!

**Review by paulineribeyre — APPROVED** [review]

**Pull request merged**